### DISCUSSION TITLE
First load of a View is slow

### DISCUSSION THREAD

**Discussion by MAUIoxo** [roubachof/Sharpnado.Tabs#137] [Q&A]
I use the Sharpnado Tabs with a `ViewSwitcher`.

My `View` I am loading is not complicated I would say. Some Background `Image`, some `Frames`, `CollectionView`. … and it also does not have too many elements to load. The `CollectionView` has item templates, but only 3-4 elements to load. My expectation is it should load and be visible very quick...

When this view is loaded for the first time using  a `LazyView` I can see the typical delay which should be solved by using the `DelayedView`. But the `DelayedView` promo video describes a situation on startup of the app. I am loading the `View` when clicking on a tab.

Nevertheless, even when using a `DelayView`, I see an activity indicator for such a simple View and I am asking myself why? When I use `LazyView` it doesn’t load for a short time and when using `DelayedView` I see the activity indicator. Both a user does not expect and I feel it is too long until it is possible to work on that `View`. After it was loaded for the first time switching back and forth in the tabs is very smooth and quick.

I even tried and pre-loaded this `View` during startup and connected this instance via Code, in both of the above `Views`. But still, when it is loaded for the first time there is too much delay and I think as a user I don’t want to accept this and it gives the impression of a lacking app.

Is there a reason for this and any way around this? Also maybe a possibility to just set a pre-loaded `View` **visible** instead of creating it when the `ViewSwitcher` loads it?

We are not talking about seconds here, but it’s enough delay to think the app has a problem.
After it was loaded for the first time, it is absolutely no problem to come back to this tab and it feels absolutely normal.

Any ideas how to solve this?

**Comment by roubachof** [comment]
Sorry I didn't really understand first.
So the first loading of the view takes some time when using small CollectionView and some frames?
Well welcome to maui XD.
If you use the DelayView, you could display a loading view right away to your user.
OR you can use just your regular view, but your startup time will be increased by the time your view is being constructed.
Also, check in Release, cause Debug is WAYYY slower, especially on MAUI because of Xaml compilation on demand.

**Reply by MAUIoxo** [reply]
Yes, that’s right - welcome to MAUI 😅

The question more or less is, why is it like that and are there any improvements I could make (in general). Maybe tomorrow, I could also try to share some snippets of the `View` if it helps anything.

Anyway, I tried both, first of all a `DelayedView` right in the tab `ViewSwitcher` and was surprised why I saw an activity indicator on first load when I switched to the corresponding tab. Then, I tried with a `LazyView` and also creating the `View` instance in advance and just assign this.

The `LazyView` approach with prepared `View` instance wasn’t really better from a user perspective as I clicked on the tab and had to way shortly until it visually moved to that `View`.

I also tried to compile the XAML `View`, but didn’t help.

Now, I am a bit frustrated: the `DelayedView` shows up quickly, but shows the activity indicator in the `CollectionView` where she really doesn’t understand why there is an activity indicator at all showing up for 3-4 items in the `CollectionView`.

Do you know of any options like pre-creation of the `View` in advance and then just setting it `Visible=true` or so in the `ViewSwitcher` or so?

**Reply by roubachof** [reply]
To sum-up:

1. Use DelayedView if you have big Xaml views (it will load tabs on demand instead of loading all of them)
2. Use LazyView if you want to load tabs on demand
3. Use none of it, use directly the view, if you want to see right away your view and you don't experience lags or long startup time

**Reply by MAUIoxo** [reply]
Thanks for overview in your summarization. Unfortunately, it didn't help in my case :(
I also exported my SVG to a JPEG file which reduced the size tremendously and is not at 258 KB, but still too slow.
Tested **Release** deployed to my **iPhone 14 Pro Max**

Same behaviour all around:

- `DelayedView` showing `ActivityIndicator`
- `LazyView` with a delay/lag after clicking the Tab
- My own View with same delay/lag after clicking as the `LazyView`

Here is the `View` (XAML) - is it already too complicated from your point of view?

**OverviewView.xaml**:
```xaml
<?xml version="1.0" encoding="utf-8" ?>
<ContentView xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             xmlns:ios="clr-namespace:Microsoft.Maui.Controls.PlatformConfiguration.iOSSpecific;assembly=Microsoft.Maui.Controls"
             xmlns:converters="clr-namespace:Optimizer.Pages.Views.Converters"
             xmlns:dx="http://schemas.devexpress.com/maui"
             xmlns:loc="clr-namespace:LocalizationResourceManager.Maui;assembly=LocalizationResourceManager.Maui"
             xmlns:views="clr-namespace:Optimizer.Pages.Views"
             xmlns:viewModels="clr-namespace:Optimizer.ViewModels"
             xmlns:databaseModels="clr-namespace:OptimizerDatabase.Models;assembly=OptimizerDatabase"
             xmlns:controls="clr-namespace:Optimizer.Pages.Views.Controls.CustomSwipeView"
             xmlns:searchBarControl="clr-namespace:Optimizer.Pages.Views.Controls.CustomSearchBar"
             xmlns:labelControls="clr-namespace:Optimizer.Pages.Views.Controls.CustomLabel"
             x:Class="Optimizer.Pages.Views.OverviewView">

    <!--<ContentView.BindingContext>
        <viewModels:OverviewViewModel/>
    </ContentView.BindingContext>-->

    <ContentView.Resources>
        <ResourceDictionary>
            <Style x:Key="BorderStyleView" TargetType="Border">
                <Setter Property="Background" Value="{StaticResource White}" />
                <Setter Property="StrokeShape" Value="RoundRectangle 10, 10, 10, 10" />
                <Setter Property="Margin" Value="15" />
                <Setter Property="Opacity" Value="0.9" />
                <Setter Property="Padding" Value="0" />
                <Setter Property="VerticalOptions" Value="Start"/>
            </Style>
            <Style x:Key="BorderRoundedCornerStyle" TargetType="Border">
                <Setter Property="StrokeShape" Value="RoundRectangle 10, 10, 10, 10" />
            </Style>
            <Style x:Key="GroupHeaderVerticalStackLayoutStyle" TargetType="VerticalStackLayout">
                <Setter Property="Margin" Value="0, 10, 0, 0" />
            </Style>
            <Style x:Key="GroupHeaderStyle" TargetType="Label">
                <Setter Property="FontSize" Value="16"/>
                <Setter Property="FontAttributes" Value="Bold"/>
                <Setter Property="TextColor" Value="{StaticResource DarkOrange1}" />
                <Setter Property="Margin" Value="0, 10, 0, 0" />
            </Style>
            <Style x:Key="GroupHeaderUnderlineStyle" TargetType="BoxView">
                <Setter Property="HeightRequest" Value="0.5" />
                <Setter Property="Color" Value="{StaticResource Gray200}" />
                <Setter Property="Margin" Value="0, 5" />
            </Style>
            <Style x:Key="SelectionDetailsLabelStyle" TargetType="Label">
                <Setter Property="FontSize" Value="11" />
                <Setter Property="TextColor" Value="{StaticResource Gray900}"/>
            </Style>
            <Style x:Key="SwipeItemStyle" TargetType="Label">
                <Setter Property="FontSize" Value="14"/>
                <Setter Property="HorizontalOptions" Value="FillAndExpand"/>
                <Setter Property="VerticalOptions" Value="FillAndExpand"/>
                <Setter Property="HorizontalTextAlignment" Value="Center"/>
                <Setter Property="VerticalTextAlignment" Value="Center"/>
            </Style>

            <converters:UpperCaseConverter x:Key="GroupNameUpperCaseConverter"/>
        </ResourceDictionary>
    </ContentView.Resources>

    
    <Grid VerticalOptions="StartAndExpand">

        <!-- Can't use the BackgroundImageView here since there is a Bug that draws the image too big and we have a gray stipe at the bottom -->
        <Image x:Name="BackgroundImage" Source="cob_gradient.jpg" VerticalOptions="StartAndExpand" HorizontalOptions="FillAndExpand" Aspect="AspectFill" Margin="{OnPlatform Android='0, -40, 0, 0', Default='0'}"/>

        <Border x:Name="sOverviewBorder" Grid.Row="0" Style="{StaticResource BorderStyleView}" VerticalOptions="StartAndExpand">

            <Grid RowDefinitions="Auto, *">

                <!--Search...-->
                <searchBarControl:SearchBarLang Grid.Row="0" ios:SearchBar.SearchBarStyle="Minimal" Text="{Binding SearchText}" Placeholder="{loc:Translate Overview_SearchItemPlaceholder}" CancelButtonText="{loc:Translate CancelButtonText}"/>

                <!-- CollectionView: Margin to get the ScrollBar to the right and closer to the Border -->
                <dx:DXCollectionView Grid.Row="1" ItemsSource="{Binding AvailableItems, Mode=TwoWay}" AllowGroupCollapse="False" FilterExpression="{Binding SearchFilterExpression}" Margin="10, 5, 10, 15">

                    <!-- Group Description: FieldName used to group the items and Order in which they are grouped -->
                    <dx:DXCollectionView.GroupDescription>
                        <dx:GroupDescription FieldName="Name" GroupInterval="Alphabetical"/>
                    </dx:DXCollectionView.GroupDescription>
                    
                    <!-- Group Header: to group items -->
                    <dx:DXCollectionView.GroupHeaderTemplate>
                        <DataTemplate>
                            <VerticalStackLayout>
                                <Label Text="{Binding Value, StringFormat=' {0}'}" Style="{StaticResource GroupHeaderStyle}" />
                                <BoxView Style="{StaticResource GroupHeaderUnderlineStyle}"/>
                            </VerticalStackLayout>
                        </DataTemplate>
                    </dx:DXCollectionView.GroupHeaderTemplate>

                    <!-- Group Items -->
                    <dx:DXCollectionView.ItemTemplate>
                        <DataTemplate x:DataType="{x:Type databaseModels:}">

                            <controls:CustomSwipeView SwipeViewOpenFlag="{Binding SwipeViewOpenFlag, Source={RelativeSource AncestorType={x:Type viewModels:OverviewViewModel}}, Mode=TwoWay}">

                                <Grid Padding="0, 5">

                                    <!-- Rectangle with rounded corners. The padding is used for the distance to the Name, Brand ... -->
                                    <Border Padding="20, 15, 20, 10" Style="{StaticResource BorderRoundedCornerStyle}">
                                        <StackLayout>

                                            <!--  Name -->
                                            <Label Text="{Binding Name}" FontSize="13" FontAttributes="Bold" />


                                            <!--  Brand -->
                                            <Label Text="{Binding Brand}" FontSize="12" FontAttributes="Bold" TextColor="{StaticResource Gray500}" Margin="0, 5, 0, 0"/>


                                            <!-- Bottom Row -->
                                            <HorizontalStackLayout Margin="0, 15, 0, 0">
                                                <HorizontalStackLayout HorizontalOptions="StartAndExpand">
                                                    <Image Source="orange1.png" HeightRequest="12" WidthRequest="12" Margin="0, 1, 5, 5"/>
                                                    <Label Text="{loc:Translate LabelStringA, StringFormat='{0}:'}" Style="{StaticResource SelectionDetailsLabelStyle}"/>
                                                    <labelControls:CultureBasedNumberLabel Text="{Binding A, Mode=OneWay, StringFormat='  {0:0.###}'}" Style="{StaticResource SelectionDetailsLabelStyle}" />
                                                </HorizontalStackLayout>

                                                <HorizontalStackLayout HorizontalOptions="CenterAndExpand" Margin="5, 0, 0, 0">
                                                    <Image Source="orange2.png" HeightRequest="20" WidthRequest="20" Margin="0, -2, 0, 0" />
                                                    <Label Text="{loc:Translate LabelStringB, StringFormat='{0}:'}" Style="{StaticResource SelectionDetailsLabelStyle}"/>
                                                    <labelControls:CultureBasedNumberLabel Text="{Binding B, Mode=OneWay, StringFormat='  {0:0.###}'}" Style="{StaticResource SelectionDetailsLabelStyle}" />
                                                </HorizontalStackLayout>

                                                <HorizontalStackLayout HorizontalOptions="EndAndExpand" Margin="5, 0, 0, 0">
                                                    <Image Source="orange3.png" HeightRequest="16" WidthRequest="16" Margin="5, -1, 3, 2" />
                                                    <Label Text="{loc:Translate LabelStringC, StringFormat='{0}:'}" Style="{StaticResource SelectionDetailsLabelStyle}"/>
                                                    <labelControls:CultureBasedNumberLabel Text="{Binding C, Mode=OneWay, StringFormat='  {0:0.###}'}" Style="{StaticResource SelectionDetailsLabelStyle}" />
                                                </HorizontalStackLayout>
                                            </HorizontalStackLayout>

                                        </StackLayout>
                                    </Border>

                                </Grid>

                                <SwipeView.RightItems>
                                    <SwipeItems>
                                        <SwipeItemView>

                                            <Grid RowDefinitions="*" ColumnDefinitions="*, *" WidthRequest="160" HeightRequest="100" ColumnSpacing="0">
                                                <Label Grid.Column="0" Text="{loc:Translate sOverview_SwipeItemEditItem}" BackgroundColor="LightGrey" Style="{StaticResource SwipeItemStyle}">
                                                    <Label.GestureRecognizers>
                                                        <TapGestureRecognizer Command="{Binding Source={RelativeSource AncestorType={x:Type viewModels:OverviewViewModel}}, Path=EditItemCommand}" CommandParameter="{Binding .}"/>
                                                    </Label.GestureRecognizers>
                                                </Label>

                                                <Label Grid.Column="1" Text="{loc:Translate sOverview_SwipeItemDeleteItem}" BackgroundColor="{StaticResource DarkOrange1}" TextColor="White" Style="{StaticResource SwipeItemStyle}">
                                                    <Label.GestureRecognizers>
                                                        <TapGestureRecognizer Command="{Binding Source={RelativeSource AncestorType={x:Type viewModels:OverviewViewModel}}, Path=DeleteItemCommand}" CommandParameter="{Binding .}"/>
                                                    </Label.GestureRecognizers>
                                                </Label>
                                            </Grid>

                                        </SwipeItemView>
                                    </SwipeItems>
                                </SwipeView.RightItems>

                            </controls:CustomSwipeView>

                        </DataTemplate>
                    </dx:DXCollectionView.ItemTemplate>

                </dx:DXCollectionView>
                
            </Grid>
            
        </Border>

    </Grid>

</ContentView>
```

**OverviewView.xaml.cs**:
```cs
namespace Optimizer.Pages.Views;

[XamlCompilation(XamlCompilationOptions.Compile)]
public partial class OverviewView : ContentView
{
	public OverviewView()
	{
		InitializeComponent();
	}

    protected override void OnSizeAllocated(double width, double height)
    {
        base.OnSizeAllocated(width, height);

        if (height < 0)
        {
            return;
        }

        if (height > width)
        {
            BackgroundImage.HeightRequest = height * 0.7;
        }
        else
        {
            BackgroundImage.HeightRequest = height;
        }

        OverviewBorder.HeightRequest = height;
    }
}
```

**Comment by roubachof** [comment]
I still don't understand what is your issue :)
Could you make the video?

**Reply by MAUIoxo** [reply]
After the **first click** on "Tab 2" it takes some time until the content of the View is visible. For a user, this timespan feels too long after a click not seeing anything and it feels like the App is not reacting or lagging at this point:

![Overview_Lagging](https://github.com/user-attachments/assets/2a9dc84c-c53a-4cde-b221-204a7d8f37d4)

After the **second click** on "Tab 2" when switching between "Tab 1" and "Tab 2" back and forth it feels way smoother reaching reaching the view of "Tab 2". The timespan is short enough so that a user doesn't get the impression anymore that the App is lagging:

![Overview_AfterInitialClick](https://github.com/user-attachments/assets/45d91b94-a9aa-45e6-b1d2-94337219e973)

**Reply by roubachof** [reply]
Yes, it's the first build of the View, it's normal on MAUI when you first construct a big View (and ViewB is a big view), you better use DelayedView with an ActivityIndicator (UseActivityIndicator="true"); so as soon as you click on B, it will select the B tab instead of lagging

**Reply by MAUIoxo** [reply]
You are right, obviously. But, I am confused that this can already be considered a "big view" - did not expect it to be big already and it also was compiled.

Yes, a `DelayedView` might help a bit, but to be honest an Activity Indicator is also strange for the user when ends up in seeing just 3-4 items in the view then. At least **I wouldn't** expect it needs an activity indicator for loading 3-4 items. And the end result, **a feeling of a lagging App** still persists in this case.

Don't get me wrong - it's not your fault or so. I am just a bit surprised that MAUI is like that with this scenario as a background and wondered if there are any tricks available to work around this.

**Reply by roubachof** [reply]
well welcome to MAUI XD
But test it first in RELEASE, its VERY IMPORTANT, cause DEBUG is super slow. So Maybe in RELEASE you won't experience any delay

**Reply by MAUIoxo** [reply]
I always also deploy it on my iPhone (Release) to see it in real life action. There, I felt like it is lagging. Switching after the first initialization feels natural and as it should be.

Do you think there is a way to load the content in a piece of Platform Code (iOS/Android) which could then be faster? I still am surprised that even when I created the instance of the view beforehand and assigned it that it is still slow, which doesn’t make sense at all.

**Reply by roubachof** [reply]
It's strange though that it's laggy on iOS release... normally it's more an Android issue.
But yeah I created the DelayedView exactly for that the lag when clicking a tab showing a big view...
Constructing the un beforehand won't change a thing cause all the layout calculations are done when you attach the view to its parent the first time.
